Source/WebKit/WebProcess/Network/WorkerWebTransportSendStreamSink.cpp

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright (C) 2023 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "WorkerWebTransportSendStreamSink.h"
+
+#include <wtf/RunLoop.h>
+
+namespace WebKit {
+
+Ref<WorkerWebTransportSendStreamSink> WorkerWebTransportSendStreamSink::create(WebCore::ScriptExecutionContextIdentifier identifier)
+{
+    return adoptRef(*new WorkerWebTransportSendStreamSink(identifier));
+}
+
+WorkerWebTransportSendStreamSink::~WorkerWebTransportSendStreamSink()
+{
+    ASSERT(!RunLoop::isMain());
+}
+
+WorkerWebTransportSendStreamSink::WorkerWebTransportSendStreamSink(WebCore::ScriptExecutionContextIdentifier identifier)
+    : m_scriptExecutionContextIdentifier(identifier)
+{
+    ASSERT(!RunLoop::isMain());
+}
+
+void WorkerWebTransportSendStreamSink::write(WebCore::ScriptExecutionContext& context, JSC::JSValue value, WebCore::DOMPromiseDeferred<void>&& promise)
+{
+    ASSERT(!RunLoop::isMain());
+
+    if (!context.globalObject())
+        return promise.reject(WebCore::Exception { WebCore::InvalidStateError });
+    auto& globalObject = *JSC::jsCast<WebCore::JSDOMGlobalObject*>(context.globalObject());
+    auto scope = DECLARE_THROW_SCOPE(globalObject.vm());
+    auto arrayBufferOrView = convert<WebCore::IDLUnion<WebCore::IDLArrayBuffer, WebCore::IDLArrayBufferView>>(globalObject, value);
+    if (scope.exception())
+        return promise.settle(WebCore::Exception { WebCore::ExistingExceptionError });
+
+    auto bytes = WTF::switchOn(arrayBufferOrView, [&](auto& arrayBufferOrView) {
+        return Vector<uint8_t> { std::span<const uint8_t> { static_cast<const uint8_t*>(arrayBufferOrView->data()), arrayBufferOrView->byteLength() } };
+    });
```

Comment:
A lot of this code will be the same in window and worker.
It might be best to reuse the same code as much as possible.
Ideally, the implementation (worker or not) would just be to make some IPC from either worker or main threads.
This will reduce copies and be faster as well, which is important for WebTransport realtime use cases.